qualtran/_infra/data_types.py

```diff
@@ -521,46 +521,120 @@ class QFxp(QDType):
 
+    def __attrs_post_init__(self):
+        if not is_symbolic(self.num_qubits) and self.num_qubits == 1 and self.signed:
+            raise ValueError("num_qubits must be > 1.")
+        if not is_symbolic(self.bitsize) and not is_symbolic(self.num_frac):
+            if self.signed and self.bitsize == self.num_frac:
+                raise ValueError("num_frac must be less than bitsize if the QFxp is signed.")
+            if self.bitsize < self.num_frac:
+                raise ValueError("bitsize must be >= num_frac.")
+
     @property
     def num_qubits(self):
         return self.bitsize
 
     @property
     def num_int(self) -> SymbolicInt:
-        return self.bitsize - self.num_frac - int(self.signed)
+        return self.bitsize - self.num_frac
+
+    def is_symbolic(self) -> bool:
+        return is_symbolic(self.bitsize, self.num_frac)
 
     @property
-    def fxp_dtype_str(self) -> str:
-        return f'fxp-{"us"[self.signed]}{self.bitsize}/{self.num_frac}'
+    def _int_qdtype(self) -> Union[QUInt, QInt]:
+        """The corresponding integer type used to represent raw values of this type.
+
+        This raw integer value is used in the classical simulator to represent values
+        of QFxp registers.
+
+        For example, QFxp(6, 2) has 2 int bits and 4 frac bits, and the corresponding
+        int type is QUInt(6). So a true classical value of `10.0011` will have a raw
+        integer representation of `100011`.
+        """
+        return QInt(self.bitsize) if self.signed else QUInt(self.bitsize)
+
+    def get_classical_domain(self) -> Iterable[int]:
+        yield from self._int_qdtype.get_classical_domain()
+
+    def to_bits(self, x) -> List[int]:
+        return self._int_qdtype.to_bits(x)
+
+    def from_bits(self, bits: Sequence[int]):
+        return self._int_qdtype.from_bits(bits)
+
+    def assert_valid_classical_val(self, val: int, debug_str: str = 'val'):
+        self._int_qdtype.assert_valid_classical_val(val, debug_str)
+
+    def to_fixed_width_int(self, x: Union[float, Fxp], *, require_exact: bool = False) -> int:
+        """Returns the interpretation of the binary representation of `x` as an integer."""
+        bits = self._fxp_to_bits(x, require_exact=require_exact)
+        return self._int_qdtype.from_bits(bits)
+
+    def __str__(self):
+        if self.signed:
+            return f'QFxp({self.bitsize}, {self.num_frac}, True)'
+        else:
+            return f'QFxp({self.bitsize}, {self.num_frac})'
+
+    # Experimental `fxpmath.Fxp` support.
+    # This support is currently experimental, and does not hook into the classical
+    # simulator protocol. Once the library choice for fixed-point classical real
+    # values is finalized, the code will be updated to use the new functionality
+    # instead of delegating to raw integer values (see above).
 
     @property
-    def _fxp_dtype(self) -> Fxp:
-        return Fxp(None, dtype=self.fxp_dtype_str)
+    def fxp_dtype_template(self) -> Fxp:
```

Comment:
why are we using the word "template" here? because it doesn't actually have a value associated with it?

```
"""Prepare an empty `fxpmath.Fxp` value container for experimental fixed point support.

This constructs a `Fxp` object with no value. To assign the returned object a fixed point value, 
use ... [idk, exaplain how to use it].

Fxp support is experimental and doesn't hook into the classical simulator protocol etc etc

This corresponds to the Fxp constructor arguments:
 - op sizing, etc etc
"""
```

Reply:
I'd also make this a method instead of a property

Reply:
I used the term from the fxpmath readme: https://github.com/francof2a/fxpmath#:~:text=It%20is%20a%20good%20idea%20create%20Fxp%20objects%20like%20template%3A

This is used to type-cast Fxp values, instead of being an empty container, hence a property.

Reply:
I've updated the docstring to explain how to use this function

Reply:
you're never going to want to plumb through any of the options?